### ISSUE TITLE
could i call this api lws_write in libwebsocket in multithread? i will record the wsi variable, and is that ok i send the data in other thread other than the protocol callback function?

### ISSUE THREAD

**Issue by pppaulpeter** [warmcat/libwebsockets#3240]
Hi Andy,
            
 this is the answer i got from chatgpt4o：

> ‘’‘No, you should not call lws_write() directly from multiple threads in libwebsockets (LWS). The library is not thread-safe in this way, and lws_write() must only be called within the context of the libwebsockets protocol callback function, typically when it is triggered by a LWS_CALLBACK_SERVER_WRITEABLE or LWS_CALLBACK_CLIENT_WRITEABLE event.
> 
> Reasons Why You Should Not Call lws_write() from Multiple Threads:
> Thread-Safety: lws_write() is not thread-safe, meaning it should not be invoked from multiple threads or from outside the main LWS event loop.
> Internal LWS State: LWS maintains internal state and buffers that are only safe to access within the callback, where the state machine controls data flow.’‘’

but according to my code which i have implemented the multithread lws_write() calling , it seems fine. could you shed some light on this? is the multithread calling lws_write() forbidden? may i know the reason?

before you have post this in another issue：

>   There's no difference in approach for this.  The callbacks you use to collect and maintain a list of active clients are slightly different for server is all.
> 
> You can keep state about the connection to the client in the wsi pss (user allocation) somewhere, if there are multiple threads it's in a composed struct that the other thread can also understand and protected by a mutex in there typically.  You can iterate through the alive wsi in the CANCELLED callback seeing if they are in a state where they have anything they want to do, eg, didn't catch up with a ringbuffer of events yet on that particular connection.  Then you are in a good, safe place to ask for writeable callbacks on any wsi that want to write something.  And actually write it again according to the state you find in the WRITEABLE callback for the wsi when that comes.
> 
> _Originally posted by @lws-team in https://github.com/warmcat/libwebsockets/issues/2936#issuecomment-1647383407_

**Comment by lws-team** [comment]
1) You can't throw data randomly at a remote peer over the network.  If it has problems transferring to it, data will pile up until it fills the tcp window, at which point it's not even trying to accept new data for send.  Instead, in order to adapt the rate of sending to the rate that it can accept, you must indicate you would like to write, but defer sending anything new until there is space in the tcp window.  The WRITEABLE callback is the indication that you are able to send something now on that connection, and you can only call lws_write() from there.  Slow connections come up with WRITEABLE slowly and fast connections come up with them immediately, so everything adapts to what is actually possible.

2) Lws apis are not threadsafe.  It's based around a single-threaded event loop, and takes advantage of that to be safe from anything else changing while it's servicing one socket.  #2396 is about using it with multiple worker threads from a single lws thread for socket comms, ie, all writes are done from the lws thread only as usual.

**Comment by pppaulpeter** [comment]
> 1. You can't throw data randomly at a remote peer over the network.  If it has problems transferring to it, data will pile up until it fills the tcp window, at which point it's not even trying to accept new data for send.  Instead, in order to adapt the rate of sending to the rate that it can accept, you must indicate you would like to write, but defer sending anything new until there is space in the tcp window.  The WRITEABLE callback is the indication that you are able to send something now on that connection, and you can only call lws_write() from there.  Slow connections come up with WRITEABLE slowly and fast connections come up with them immediately, so everything adapts to what is actually possible.
> 2. Lws apis are not threadsafe.  It's based around a single-threaded event loop, and takes advantage of that to be safe from anything else changing while it's servicing one socket.  [Can sending and receiving messages only be in callback functions? #2936](https://github.com/warmcat/libwebsockets/issues/2936) is about using it with multiple worker threads from a single lws thread for socket comms, ie, all writes are done from the lws thread only as usual.

ok, thanks, could the worker thread calls lws_callback_on_writable(wsi) to indicate that the WebSocket instance (wsi) has data ready to be sent?lws_callback_on_writable is thread safe as the chatgpt suggest? :）

**Comment by lws-team** [comment]
Whatever chatgpt says or doesn't say is not my problem to justify.  It sounds like it's your problem for asking it.

You cannot call lws_callback_on_writable() from any thread other than  the lws event thread.  Because modifying the fd wait list from another thread does not interrupt any ongoing wait on the lws thread.

**Comment by pppaulpeter** [comment]
> Whatever chatgpt says or doesn't say is not my problem to justify. It sounds like it's your problem for asking it.
> 
> You cannot call lws_callback_on_writable() from any thread other than the lws event thread. Because modifying the fd wait list from another thread does not interrupt any ongoing wait on the lws thread.

yes, it is my problem.
just to be clear, i call  lws_callback_on_writable()  in the protocol callback(the event loop) and there was still 14ms delay before going to the case LWS_CALLBACK_SERVER_WRITEABLE，may i know is that normal?

**Comment by lws-team** [comment]
The idea of separating the request and the event that you can write is that the tcp window will control when or if the callback happens.  If the tcp window on the connection is filled and no ACKs come to draw it down to allow to send more, you will never get another WRITEABLE callback after the window is full.  Or, if the link has a huge bandwidth and the data is being used on the peer side, you will get the callback immediately (which is platform dependent, maybe 50us to complete the event loop).  This is how lws adapts your sending rate to the bandwidth of the link, without any waste for buffering usendable things in RAM.

If you try to send something bigger than can be sent by the kernel (this is decided kernelside by kernel memory situation and the socket tcp window state), lws will buffer the unsendable part in heap and conceal the WRITEABLE from your code until it has finished sending the remainder using them.  So if you pointlessly send 10MB or whatever in one lws_write(), you will be made to wait until it has all gone, and 10MB of heap will be wasted.

So it can be perfectly normal for tcp to tell you to stop sending for 14ms depending on what you're doing and your connection.

Have doubts?  Can't understand it?  I don't care :-)

**Comment by pppaulpeter** [comment]
> The idea of separating the request and the event that you can write is that the tcp window will control when or if the callback happens. If the tcp window on the connection is filled and no ACKs come to draw it down to allow to send more, you will never get another WRITEABLE callback after the window is full. Or, if the link has a huge bandwidth and the data is being used on the peer side, you will get the callback immediately (which is platform dependent, maybe 50us to complete the event loop). This is how lws adapts your sending rate to the bandwidth of the link, without any waste for buffering usendable things in RAM.
> 
> If you try to send something bigger than can be sent by the kernel (this is decided kernelside by kernel memory situation and the socket tcp window state), lws will buffer the unsendable part in heap and conceal the WRITEABLE from your code until it has finished sending the remainder using them. So if you pointlessly send 10MB or whatever in one lws_write(), you will be made to wait until it has all gone, and 10MB of heap will be wasted.
> 
> So it can be perfectly normal for tcp to tell you to stop sending for 14ms depending on what you're doing and your connection.
> 
> Have doubts? Can't understand it? I don't care :-)

Hi andy,

thanks again for you quick reply, but here the 14 ms delay is the first time delay before i calling the lws_write(), but i can't reproduce the 14 ms delay problem in my environment either, now the delay is less then 1ms. probably it is network issue. thanks for your help!:-)
